### PR TITLE
Stop copying LogicalPlan and Exprs in `EliminateCrossJoin` (4% faster planning)

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on~
* https://github.com/apache/datafusion/pull/10410
* https://github.com/apache/datafusion/pull/10427
* https://github.com/apache/datafusion/pull/10430

## Which issue does this PR close?
Closes https://github.com/apache/datafusion/issues/10287

## Rationale for this change

Make planning faster by avoiding copying

## What changes are included in this PR?

1. Change `EliminateCrossJoin` to use TreeNode API and stop copying plans
2. Reduce `Expr` copies more
3. Improve documentation

## Are these changes tested?

Existing CI

## Are there any user-facing changes?
No functional changes,

Performance change: 4% better planning

(note this includes the improvements from https://github.com/apache/datafusion/pull/10427 as well)

<details><summary>Details</summary>
<p>

```
group                                         eliminate_cross_join                   main
-----                                         --------------------                   ----
logical_aggregate_with_join                   1.00  1214.6±15.54µs        ? ?/sec    1.00  1218.8±14.06µs        ? ?/sec
logical_plan_tpcds_all                        1.01    161.6±1.93ms        ? ?/sec    1.00    159.4±1.84ms        ? ?/sec
logical_plan_tpch_all                         1.02     17.2±0.19ms        ? ?/sec    1.00     16.9±0.16ms        ? ?/sec
logical_select_all_from_1000                  1.00     18.0±0.08ms        ? ?/sec    1.05     18.8±0.09ms        ? ?/sec
logical_select_one_from_700                   1.00   822.4±15.44µs        ? ?/sec    1.00   820.4±44.09µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.01    765.5±9.28µs        ? ?/sec    1.00   759.0±13.89µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.01    751.8±9.19µs        ? ?/sec    1.00   741.7±23.29µs        ? ?/sec
physical_plan_tpcds_all                       1.00   1310.5±5.75ms        ? ?/sec    1.04   1357.6±7.25ms        ? ?/sec
physical_plan_tpch_all                        1.00     90.1±0.88ms        ? ?/sec    1.04     94.1±0.82ms        ? ?/sec
physical_plan_tpch_q1                         1.00      3.9±0.05ms        ? ?/sec    1.33      5.2±0.07ms        ? ?/sec
physical_plan_tpch_q10                        1.00      4.3±0.06ms        ? ?/sec    1.03      4.4±0.07ms        ? ?/sec
physical_plan_tpch_q11                        1.00      3.9±0.06ms        ? ?/sec    1.03      4.0±0.05ms        ? ?/sec
physical_plan_tpch_q12                        1.00      3.0±0.05ms        ? ?/sec    1.02      3.1±0.07ms        ? ?/sec
physical_plan_tpch_q13                        1.00      2.1±0.03ms        ? ?/sec    1.00      2.2±0.04ms        ? ?/sec
physical_plan_tpch_q14                        1.00      2.5±0.04ms        ? ?/sec    1.10      2.8±0.04ms        ? ?/sec
physical_plan_tpch_q16                        1.00      3.6±0.04ms        ? ?/sec    1.07      3.9±0.04ms        ? ?/sec
physical_plan_tpch_q17                        1.00      3.5±0.05ms        ? ?/sec    1.04      3.7±0.05ms        ? ?/sec
physical_plan_tpch_q18                        1.00      3.9±0.07ms        ? ?/sec    1.03      4.1±0.04ms        ? ?/sec
physical_plan_tpch_q19                        1.00      5.9±0.05ms        ? ?/sec    1.08      6.4±0.05ms        ? ?/sec
physical_plan_tpch_q2                         1.00      7.7±0.06ms        ? ?/sec    1.03      7.9±0.05ms        ? ?/sec
physical_plan_tpch_q20                        1.00      4.6±0.06ms        ? ?/sec    1.02      4.7±0.06ms        ? ?/sec
physical_plan_tpch_q21                        1.00      6.2±0.08ms        ? ?/sec    1.02      6.3±0.05ms        ? ?/sec
physical_plan_tpch_q22                        1.00      3.4±0.06ms        ? ?/sec    1.04      3.5±0.05ms        ? ?/sec
physical_plan_tpch_q3                         1.00      3.1±0.04ms        ? ?/sec    1.07      3.3±0.04ms        ? ?/sec
physical_plan_tpch_q4                         1.00      2.3±0.03ms        ? ?/sec    1.04      2.4±0.06ms        ? ?/sec
physical_plan_tpch_q5                         1.00      4.4±0.06ms        ? ?/sec    1.05      4.7±0.05ms        ? ?/sec
physical_plan_tpch_q6                         1.00  1507.9±41.94µs        ? ?/sec    1.06  1598.3±24.25µs        ? ?/sec
physical_plan_tpch_q7                         1.00      5.7±0.06ms        ? ?/sec    1.03      5.8±0.07ms        ? ?/sec
physical_plan_tpch_q8                         1.00      7.3±0.05ms        ? ?/sec    1.03      7.5±0.07ms        ? ?/sec
physical_plan_tpch_q9                         1.00      5.6±0.07ms        ? ?/sec    1.02      5.7±0.07ms        ? ?/sec
physical_select_all_from_1000                 1.00     58.7±0.25ms        ? ?/sec    1.05     61.5±1.01ms        ? ?/sec
physical_select_one_from_700                  1.00      3.7±0.03ms        ? ?/sec    1.02      3.7±0.05ms        ? ?/sec
```
</p>
</details>

### CODE REVIEW

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 1]
```diff
@@ -237,7 +324,7 @@ fn find_inner_join(
             )?);
 
             return Ok(LogicalPlan::Join(Join {
-                left: Arc::new(left_input.clone()),
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs:228]: here is one example where the plans were cloned

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 2]
```diff
@@ -144,49 +191,89 @@ impl OptimizerRule for EliminateCrossJoin {
...
                 ..
             })
             | LogicalPlan::CrossJoin(_) => {
-                if !try_flatten_join_inputs(child, possible_join_keys, all_inputs)? {
-                    return Ok(false);
+                if !can_flatten_join_inputs(child) {
+                    return false;
                 }
             }
-            _ => all_inputs.push(child.clone()),
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs]: here is an example of the clones that are removed

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 3]
```diff
@@ -259,7 +346,7 @@ fn find_inner_join(
     )?);
 
     Ok(LogicalPlan::CrossJoin(CrossJoin {
-        left: Arc::new(left_input.clone()),
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs:247]: likewise here is another removed clone

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 4]
```diff
@@ -39,102 +41,147 @@ impl EliminateCrossJoin {
...
         }
 
         let Some(predicate) = parent_predicate else {
-            return Ok(Some(left));
+            return Ok(Transformed::yes(left));
         };
 
         // If there are no join keys then do nothing:
         if all_join_keys.is_empty() {
-            Filter::try_new(predicate.clone(), Arc::new(left))
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs:130]: also removed several expr clones

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 5]
```diff
@@ -39,102 +41,147 @@ impl EliminateCrossJoin {
...
-                            &mut possible_join_keys,
-                        );
-                        Some(&filter.predicate)
-                    }
-                    _ => {
-                        return utils::optimize_children(self, plan, config);
-                    }
-                }
+
+        let can_flatten_inputs = can_flatten_join_inputs(&plan);
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs:101]: this took a fair amount of finagling to split up the cases where the rewrite will happen and thus we should destructure the Filter and when not. I believe the logic is all the same, but the code needed to be reorganized

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 6]
```diff
@@ -39,102 +41,147 @@ impl EliminateCrossJoin {
...
 }
 
-/// Attempt to reorder join to eliminate cross joins to inner joins.
-/// for queries:
-/// 'select ... from a, b where a.x = b.y and b.xx = 100;'
-/// 'select ... from a, b where (a.x = b.y and b.xx = 100) or (a.x = b.y and b.xx = 200);'
-/// 'select ... from a, b, c where (a.x = b.y and b.xx = 100 and a.z = c.z)
-/// or (a.x = b.y and b.xx = 200 and a.z=c.z);'
-/// 'select ... from a, b where a.x > b.y'
+/// Eliminate cross joins by rewriting them to inner joins when possible.
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs:44]: While I was in here and had it all in my head, I updated the documentation to explain how things worked

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 7]
```diff
@@ -39,102 +41,147 @@ impl EliminateCrossJoin {
...
+                    join_type: JoinType::Inner,
+                    ..
+                }) | LogicalPlan::CrossJoin(_)
+            );
+
+            if !rewriteable {
+                return rewrite_children(self, LogicalPlan::Filter(filter), config);
             }
+
+            if !can_flatten_join_inputs(&filter.input) {
```
Comment (comphead) [inline datafusion/optimizer/src/eliminate_cross_join.rs:117]: this condition can go earlier to save some cpu?

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 8]
```diff
@@ -144,49 +191,89 @@ impl OptimizerRule for EliminateCrossJoin {
...
-    let children = match plan {
+) -> Result<()> {
+    match plan {
         LogicalPlan::Join(join) if join.join_type == JoinType::Inner => {
+            // checked in can_flatten_join_inputs
             if join.filter.is_some() {
-                // The filter of inner join will lost, skip this rule.
-                // issue: https://github.com/apache/datafusion/issues/4844
-                return Ok(false);
+                return internal_err!(
```
Comment (comphead) [inline datafusion/optimizer/src/eliminate_cross_join.rs:224]: 👍

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 9]
```diff
@@ -39,102 +41,147 @@ impl EliminateCrossJoin {
...
-            return Ok(Some(left));
+            return Ok(Transformed::yes(left));
         };
 
         // If there are no join keys then do nothing:
         if all_join_keys.is_empty() {
-            Filter::try_new(predicate.clone(), Arc::new(left))
-                .map(|f| Some(LogicalPlan::Filter(f)))
+            Filter::try_new(predicate, Arc::new(left))
+                .map(LogicalPlan::Filter)
```
Comment (comphead) [inline datafusion/optimizer/src/eliminate_cross_join.rs:176]: can we do it in 1 map iteration ?

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 10]
```diff
@@ -39,102 +41,147 @@ impl EliminateCrossJoin {
...
+                .map(Transformed::yes)
         } else {
             // Remove join expressions from filter:
-            match remove_join_expressions(predicate.clone(), &all_join_keys) {
+            match remove_join_expressions(predicate, &all_join_keys) {
                 Some(filter_expr) => Filter::try_new(filter_expr, Arc::new(left))
-                    .map(|f| Some(LogicalPlan::Filter(f))),
-                _ => Ok(Some(left)),
+                    .map(LogicalPlan::Filter)
```
Comment (comphead) [inline datafusion/optimizer/src/eliminate_cross_join.rs:182]: would be nice to have 1 map iteration

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 11]
```diff
@@ -39,102 +41,147 @@ impl EliminateCrossJoin {
...
+/// ```text
+/// Filter(a.x = b.y AND b.xx = 100)
+///  CrossJoin
+///   TableScan a
+///   TableScan b
+/// ```
+///
+/// After the rule is applied, the plan will look like this:
+/// ```text
+/// Filter(b.xx = 100)
```
Comment (comphead) [inline datafusion/optimizer/src/eliminate_cross_join.rs:62]: I hope the pushdown filter predicate rule comes after this rule

**Review by comphead — APPROVED** [review]
Thanks @alamb I'd say it is great

File: datafusion/optimizer/src/eliminate_cross_join.rs (hunk 11 above)
Comment (alamb) [reply datafusion/optimizer/src/eliminate_cross_join.rs:62]: Yes absolutely 

I double checked and `PushDownFilter` does indeed run after this pass: https://github.com/apache/datafusion/blob/9cc981b06115ee40b53384c287689ce0e07950bc/datafusion/optimizer/src/optimizer.rs#L248-L259

File: datafusion/optimizer/src/eliminate_cross_join.rs (hunk 7 above)
Comment (alamb) [reply datafusion/optimizer/src/eliminate_cross_join.rs:117]: I think the call to `rewrite_children` needs to happen to apply the rule recursively before proceeding

I am not quite sure what else we could save. 

I tried to make this clearer in 030444dc3 with comments and moving another definition of `can_flatten_join_inputs` closer to where it was used.

File: datafusion/optimizer/src/eliminate_cross_join.rs (hunk 9 above)
Comment (alamb) [reply datafusion/optimizer/src/eliminate_cross_join.rs]: Done in  cb8c1f146

File: datafusion/optimizer/src/eliminate_cross_join.rs (hunk 10 above)
Comment (alamb) [reply datafusion/optimizer/src/eliminate_cross_join.rs]: Done in cb8c1f146

**Comment by alamb** [comment]
Thank you for the review @comphead  -- I think I addressed all your comments

**Pull request merged**